File: C/Recap/ScatterGather/scatter.c
#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>

#define TAG 100

int main(int argc, char ** argv)
{

  int rank, nproc, recv, irank;
  int *values;

  MPI_Init(&argc, &argv);
  MPI_Comm_size(MPI_COMM_WORLD, &nproc);
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);

  //This is using ANSI C, so a bit old fashioned
  values = (int*) malloc(sizeof(int) * nproc);

  if (rank == 0) {
    for (irank = 0; irank < nproc; ++irank) {
      printf("Please enter an integer number for processor %3d:",irank);
      fflush(stdout);
      scanf("%d", values+irank);
    }

    printf("Output\n");
    printf("------\n");
    printf("Scattering input values to other processors\n");
  }
  MPI_Barrier(MPI_COMM_WORLD);
  MPI_Scatter(values, 1, MPI_INT, &recv, 1, MPI_INT, 0,
      MPI_COMM_WORLD);
  //Can now free values
  free(values);

  printf("Rank %3d has value from scatter of %3d\n", rank, recv);

  MPI_Finalize();

}

File: Fortran/Recap/ScatterGather/scatter.f90
PROGRAM scatter

  USE mpi
  IMPLICIT NONE

  INTEGER, PARAMETER :: tag = 100

  INTEGER :: rank, recv
  INTEGER :: nproc, ierr, irank
  INTEGER, DIMENSION(:), ALLOCATABLE :: values

  CALL MPI_Init(ierr)

  CALL MPI_Comm_size(MPI_COMM_WORLD, nproc, ierr)
  CALL MPI_Comm_rank(MPI_COMM_WORLD, rank, ierr)
  ALLOCATE(values(0:nproc-1))

  IF (rank == 0) THEN
    DO irank = 0, nproc - 1
      WRITE(*,'(A I3 A)', ADVANCE='NO') &
          'Please enter an integer number for processor', irank, ":"
      READ(*,'(2I20)') values(irank)
    END DO
  END IF

  IF (rank == 0) THEN
    PRINT *,'Output'
    PRINT *,'------'
    PRINT *,'Scattering input values to other processors'
  END IF
  CALL MPI_Barrier(MPI_COMM_WORLD, ierr)
  CALL MPI_Scatter(values, 1, MPI_INTEGER, recv, 1, MPI_INTEGER, 0, &
      MPI_COMM_WORLD, ierr)
  WRITE(*,'(A I3 A I10)') 'Rank ', rank, ' has value from scatter of ', recv

  CALL MPI_Finalize(ierr)

END PROGRAM scatter
